File: java/src/main/java/toolgood/algorithm/internals/functions/string/Function_CONCATENATE.java
package toolgood.algorithm.internals.functions.string;

import java.util.List;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_N;
import toolgood.algorithm.internals.functions.NoneEngine;

public final class Function_CONCATENATE extends Function_N {
    public Function_CONCATENATE(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public String Name() {
        return "Concatenate";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, java.util.function.BiFunction<AlgorithmEngine, String, Operand> tempParameter) throws Exception {
        if (funcs.length == 0) {
            return Operand.Create("");
        }
        if (funcs.length == 1) {
            Operand a = GetText(engine, tempParameter, 0);
            if (a.IsErrorOrNone()) { return a; }
            return a;
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < funcs.length; i++) {
            Operand a = GetText(engine, tempParameter, i);
            if (a.IsErrorOrNone()) { return a; }
            sb.append(a.TextValue());
        }
        return Operand.Create(sb.toString());
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.TEXT;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        for (int i = 0; i < funcs.length; i++) {
            funcs[i].GetParameterTypes(noneEngine, result, OperandType.TEXT);
        }
    }
}

File: csharp/ToolGood.Algorithm/Internals/Functions/String/Function_CONCATENATE.cs
using System;
using System.Collections.Generic;
using System.Text;
using ToolGood.Algorithm.Enums;

namespace ToolGood.Algorithm.Internals.Functions.String
{
	internal sealed class Function_CONCATENATE : Function_N
	{
		public Function_CONCATENATE(FunctionBase[] funcs) : base(funcs)
		{
		}

		public override string Name => "Concatenate";

		public override Operand Evaluate(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			if (funcs.Length == 0) {
				return Operand.Create(string.Empty);
			}
			if (funcs.Length == 1) {
				var a = GetText(engine, tempParameter, 0);
				if (a.IsErrorOrNone) { return a; }
				return a; // 只有一个参数
			}
			var sb = new StringBuilder();
			for (int i = 0; i < funcs.Length; i++) {
				var a = GetText(engine, tempParameter, i);
				if (a.IsErrorOrNone) { return a; }
				sb.Append(a.TextValue);
			}
			return Operand.Create(sb.ToString());
		}
		public override OperandType GetResultType()
		{
			return OperandType.TEXT;
		}

		internal override void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, string op = null, string val = null)
		{
			for(int i = 0; i < funcs.Length; i++) {
				funcs[i].GetParameterTypes(noneEngine, result, OperandType.TEXT);
			}
		}
	}
}